### PR TITLE
fix(ci): select shards from the PR's own change; route tests and uncovered lines

### PR DESCRIPTION
**Pull request by ooples**
## Summary

Almost every runtime pull request has been running the **full 116-shard matrix** (~2h+) even though shard selection "worked" in every proof. This PR fixes the three causes, each reproduced on real data first.

**Headline, on real data:** #2100's exact merge commit (`ecf5a4ba8`), the exact certified map CI downloaded (run 34428489968, map sha `8decd96c`), and the exact inputs CI passed:

| | Changed files considered | Escalation reasons | Shards |
|---|---|---|---|
| **Before** (master's selector) | 22 | 25 | **116 / 116** (escalated) |
| **After** (this PR) | 4 (the PR's own) | 0 | **4 / 116** |

The "before" row isn't an estimate. Running master's selector locally on that commit reproduces CI's log line for line: the same 22 files and the same 25 reasons (log of job 103066400544).

## Why every earlier proof passed but real PRs didn't

The earlier canaries were **fresh PRs branched from current master that touched only mapped `src/` lines or non-runtime files**. That's the one shape that avoids all three causes below. #2118 "worked" only because it was non-runtime (docs + an independent workflow), which never reaches coverage selection. Real PRs are usually behind master and almost always add or edit a test, and each of those alone forces the full matrix.

## Root cause 1: the stale base (`github.event.pull_request.base.sha`)

**Before.** The Select step passed `-BaseSha ${{ github.event.pull_request.base.sha }}`. That SHA is the base branch *as it was when the PR was opened*, not what GitHub merged it onto for this run. For #2100 it was `8decd96c`, but the merge ref was built on `88364e91`, **33 commits later**. The selector's own rule, "a selection-control edit in *this PR* forces the full matrix", then fired on 16 CI files master had merged in between:

```
reason: current validation-selection control change: .github/workflows/sonarcloud.yml
reason: current validation-selection control change: tools/TestImpact/Select-Shards.ps1
... (16 total, none of them touched by #2100)
```

It feeds itself: every CI-fix PR edits `tools/TestImpact/`, so each merge pushes every *behind* PR into a full run.

**After.** The workflow passes the PR head (`github.event.pull_request.head.sha`). `Resolve-PullRequestBase` checks that the checkout is a **two-parent merge whose second parent is that head**, then uses the **first parent**, the base actually tested. A non-merge checkout, or a head that isn't the second parent, fails closed. Only the PR's own paths are selected for. Line ranges still come from the map commit, the only coordinates the map's ranges are expressed in. Two new fail-closed cases:

- The PR's own path set is empty → escalate. Scoping would otherwise make this look non-runtime.
- The PR changed a file back to exactly the map's copy, so it has no map-to-HEAD hunks → escalate. Its effect has no map coordinates.

## Root cause 2: test files are never in the coverage map

**Before.** Coverage instruments product code only. **5,931 of the map's 5,932 files are under `src/`**, so *every* test-file edit escalated (`not executed by any mapped shard: tests/.../TradingAgentLearningTests.cs`). Almost every PR adds or edits a test.

**After.** Changed test sources are routed to the shards whose `.github/test-shards.yml` filters select their tests (`Get-TestFileRoutes`):

- **Test names are read from the C#.** Comments, strings and char literals are blanked first, so a `{` in a string or `class X` in a comment can't mislead brace matching. That gives namespace, nested `Outer+Inner` chains, and `[Fact]`/`[Theory]` methods, which become the VSTest `FullyQualifiedName`.
- **Filters are parsed with VSTest's grammar** (`& | ( )`, `= != ~ !~`, `&` binding tighter) and evaluated with **three-valued logic**. A shard is skipped only when its filter is *definitely false* for every test. Whatever the parser can't know counts as "might match": casing disagreements, categories inherited from a base, and method names of inherited tests. An inherited test's unknown part is modelled as one method identifier, so it can't acquire a namespace term.
- **Shared test code follows its consumers.** Routing takes the transitive closure of test files that name the changed file's top-level types (base classes, fixtures, helpers) and unions their shards, capped at 200 files.
- **Anything reference search can't follow still escalates:** extension methods, global usings, assembly attributes, module initializers, xUnit collection definitions (bound by string name), a support file with no tests and no consumers, and tests no shard filter selects. So do **abstract classes and any type the source generator names**: `TestScaffoldGenerator` emits test classes deriving from test-project bases at *build time*, which exist in no file reference search could read.

**Measured on 250 randomly sampled real test files** (seed 20260911): **82% route, median 1 shard, mean 2.2, max 16.** The rest escalate for the conservative reasons listed above.

## Root cause 3: unexecuted changed lines

**Before.** A changed range no shard executes escalated. On #2100 all 8 such ranges were field declarations, attributes or between-method insertion points, lines coverage never records:

```
reason: changed range 60-61 is not executed by any mapped shard: src/Finance/Trading/Agents/FinancialDQNAgent.cs
```

**After.** Per your decision, such a range routes to **every shard that executes any line of the same file**. A new `.cs` file routes to the owners of its nearest mapped directory, never climbing above a two-segment root like `src/Finance`. **Still escalates:** unmapped non-C# files (`.csproj`, props, data), orphans with no mapped neighbour, and `src/AiDotNet.Generators/**`, now explicit full validation. That code runs inside the compiler, so runtime coverage never records it, yet one edit can rewrite thousands of generated tests. Before this PR it escalated only because that directory happened to be unmapped.

This is a heuristic, and the nightly selection-miss audit is its backstop. The audit now receives the audited tree's shard manifest, so it replays exactly the routing PRs get against complete matrices, and any miss revokes the map certificate (PRs then fall back to full runs).

## Every selected shard now says why

```
selected 4 of 116 shard(s)
  Integration C - Core
      because it executes src/Finance/Trading/Agents/FinancialA2CAgent.cs, whose changed lines 148-149 no shard executes
      because it executes changed lines 111-112, 245-246 of src/Finance/Trading/Agents/FinancialDQNAgent.cs
  Integration D
      because it is always run
  Integration E-G
      because it executes changed lines 123-123 of src/Finance/Trading/Agents/FinancialA2CAgent.cs
      because it runs tests affected by tests/AiDotNet.Tests/IntegrationTests/Finance/TradingAgentLearningTests.cs (its filter selects tests in this file)
  ModelFamily - Generated Layers F
      because it executes changed lines 111-112, 127-127, 227-227, 245-246 of src/Finance/Trading/Agents/FinancialDQNAgent.cs
```

## Proof

| Proof you asked for | Status |
|---|---|
| **Reproduce the failure first** | ✅ Master's selector on #2100's real merge commit reproduces CI's 25 reasons. The new end-to-end proof also builds a real behind-master repository and first asserts the **old invocation escalates**, citing master's merged control file and the unmapped test. |
| **Re-run #2100 as the real case** | ✅ Locally: 116 → 4 (table above). ⏳ In CI, after this merges: pull-request runs use the workflow from the merge ref, so #2100 benefits only once this is on master. |
| **Regression tests in the self-test** | ✅ 40+ new assertions: behind-master scoping, empty/unmappable PR path sets, owner and directory routing, filter grammar and three-valued edge cases, C# parsing traps, closure/hazard routing, and generator paths. **The new end-to-end proof fails on master's selector.** |
| **Post-merge reuse on a behind PR** | ⏳ Not in this PR: exact-tree reuse (root cause 3 of the post-merge problem) is the next PR. |

All nine impact-tooling suites pass locally under pwsh 7.6: `Select-Shards -SelfTest`, `Measure-SelectionMiss -SelfTest`, `Test-CiImpactWorkflow`, `Test-TestImpactEndToEnd`, `New-ShardMap -SelfTest`, `Select-CoverageShards -SelfTest`, `Test-CertifiedShardMap -SelfTest`, `Test-CiGateModes`, `Test-ValidationReuseModes`.

**This PR's own CI will run the full matrix. That's correct:** it edits selection-control files, which must always get complete validation.

## Files

| File | Change |
|---|---|
| `tools/TestImpact/Select-Shards.ps1` | `-PullRequestHeadSha` (merge-parent base, verified) and `-ShardManifestFile`; PR scoping; owner/directory routing; VSTest filter parser + three-valued evaluator; C# code-only lexer and test-shape reader; test routing with hazards; `routes` in the JSON output; build-time directory as full validation; self-tests. |
| `.github/workflows/sonarcloud.yml` | Select step: `PR_HEAD_SHA` replaces the event `base.sha`; writes `shard-manifest.json`; passes both to the classifier and selector. |
| `.github/workflows/test-impact-map.yml` | Nightly audit builds the audited tree's manifest and passes it through, so routing misses are measured. |
| `tools/TestImpact/Measure-SelectionMiss.ps1` | `-ShardManifestFile` passthrough. |
| `tools/TestImpact/Test-CiImpactWorkflow.ps1` | Contracts: the Select step must not read `pull_request.base.sha` or pass `-BaseSha`; must pass the head and manifest; the selector must verify both merge parents; the audit must pass the manifest. |
| `tools/TestImpact/Test-TestImpactEndToEnd.ps1` | Real-git behind-master scenario: reproduction, fix, wrong-head and non-merge fail-closed. |

## What this deliberately does not do

- **Post-merge reuse on master** (full re-run after merging a behind PR) is the next PR. Reuse still requires an exact tree match.
- **The regression-baseline lookup** (`find-test-baseline.ps1`, the retry-baseline and comparison steps) still uses the event `base.sha`. Switching it to the merge parent is correct in principle, but that lookup needs an *exact* master run, and 11 of the last 12 master runs were cancelled by the next merge. So today it would usually find no baseline at all. It belongs with the master-run work.
- **Existing coverage gap, found while building this:** statically, **~3,439 test methods in 329 files match no shard filter**, so they never run in PR *or* master CI. The largest groups are `UnitTests.NeuralNetworks` (all but four named classes), `UnitTests.Evolution.Programs`, `ContinualLearning`, `UnitTests.Agentic`, and `Generators`. `check-shard-coverage.ps1` only checks generated ModelFamily classes, so nothing reports this. Some are intentional (GPU/stress) and some are abstract bases whose tests run through derived classes, so the number needs confirming against real test discovery before anything changes. Edits to these files still escalate here, which is safe but runs 116 shards that don't include the edited tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017otuSGr3GdmvPoYLbiaWR2

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved pull request change detection using the actual PR head and merge history.
  * Improved test-impact selection by routing changed test sources to the shards that can execute them.
  * Added safer handling for unmapped files, incomplete merge checkouts, and unselectable tests.
  * Added route details to shard-selection results and logs.

* **Tests**
  * Expanded validation for behind-master pull requests, merge commits, shard routing, workflow safety, and audit consistency.

### CODE REVIEW

File: tools/TestImpact/Select-Shards.ps1 [hunk 1]
```diff
@@ -322,17 +346,64 @@ function Get-ChangedRanges {
...
+    $segments = @(([string] $Path).Replace('\', '/').Split('/'))
+    for ($depth = $segments.Count - 1; $depth -ge 2; $depth--) {
+        $directory = ($segments[0..($depth - 1)] -join '/') + '/'
+        $owners = [System.Collections.Generic.SortedSet[string]]::new([StringComparer]::Ordinal)
+        foreach ($property in $Map.files.PSObject.Properties) {
+            if (-not $property.Name.StartsWith($directory, [StringComparison]::OrdinalIgnoreCase)) { continue }
+            foreach ($occurrence in @($property.Value)) {
+                [void] $owners.Add([string] $Map.knownShards[[int] $occurrence.s])
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tools/TestImpact/Select-Shards.ps1]: Fixed in 2905be09af. Directory ownership is now indexed lazily once per Select-ImpactedShards invocation, with typed dictionary/sorted-set storage and no script-global cache. The deterministic regression measured 8 map accesses before vs 4 after for three unmapped paths; directly mapped edits still use only one access. Alternating maps in one process prove no owner leakage. I independently reviewed the implementation and reran the selector self-test and real-Git end-to-end proof.

File: tools/TestImpact/Test-CiImpactWorkflow.ps1 [hunk 2]
```diff
@@ -251,8 +251,23 @@ Assert-Contract (-not $selectStep.Contains('-AuditUnchangedMap')) `
...
+Assert-Contract ($selectStep.Contains('-ClassifyOnly `') -and
+        $selectStep.Contains('-PullRequestHeadSha $env:PR_HEAD_SHA -OutFile path-classification.json')) `
     'non-runtime classification does not use the exact PR base-to-head path set'
+Assert-Contract ($selectStep.Contains('PR_HEAD_SHA: ${{ github.event.pull_request.head.sha }}')) `
+    'the selector is not given the pull request head it must verify as the merge commit''s second parent'
+# The event base.sha is the base branch as it was when the pull request was opened. For a pull
+# request behind master it made every commit master gained since look like part of the pull
+# request: on #2100 it attributed 16 merged CI-control files to a 4-file change, escalating 116 shards.
+Assert-Contract (-not $selectStep.Contains('pull_request.base.sha') -and -not $selectStep.Contains('-BaseSha')) `
+    'pull-request selection reads the stale event base.sha instead of the merge commit''s first parent'
```
[the remark this answers is not in the document]
Comment (ooples) [reply tools/TestImpact/Test-CiImpactWorkflow.ps1:312]: Fixed in 2905be09af. The contract now parses the actual map-backed PowerShell command AST and requires its PR_HEAD_SHA argument; classifier arguments or comments cannot satisfy it. Three mutations (argument removed, wrong variable, comment-only argument) were accepted before and are rejected after. The baseline and real-Git end-to-end controls pass; a behind-master PR selects 2/3 shards, docs-only selects 0, and genuine CI controls still escalate. This regression runner is invoked by the existing end-to-end check in both workflows.

**Comment by ooples** [comment]
Timed dependency review at dce65d3263237426ac9263221e4140654830d586: zero unresolved threads. Locally ran Select-Shards.ps1 -SelfTest, Measure-SelectionMiss.ps1 -SelfTest, and Test-TestImpactEndToEnd.ps1; all passed. This does NOT make the PR merge-ready: hosted run 34596861846 failed DBN loss, StyDiff clone parity, custom-Transformer InputLayer reconstruction, and distributed-builder backend-consumption tests. Integration D received a runner shutdown signal (the log alone does not prove OOM). Vercel also reports a build-rate limit. #2183 targets DBN; #2112 describes StyDiff/COW fixes; #2185 concerns related cloning defects, but neither is established here as fixing these exact failures. No selector assertion was weakened and no blanket rerun was requested.

**Comment by ooples** [comment]
Runtime follow-up now has concrete dependency proof: #2185 at 2811e7aee2 passed the exact custom-Transformer InputLayer failure plus three clone regressions (4/4 locally). #2112 current-head hosted logs explicitly pass StyDiff clone parity (65/65 shard) and Integration D (2091 passed, 28 skipped). #2183 at 0e0c2a9f34 fixes both lazy-gradient snapshot timing and stale shard-layout caching in shared base classes; final local run passed 76/76, including the original DDP test, all five affected strategies and two-rank math checks. These fixes are in their dependency PRs, not yet merged into this selector branch. No claim that #2156 is green or merge-ready until those dependencies land and checks clear.
